### PR TITLE
LibWeb/Fetch: Implement fetch record and fetch group

### PR DESCRIPTION
**Pull request by mobounya**
Implement 'fetch records' and 'fetch group' concepts from the fech specs:
https://fetch.spec.whatwg.org/#concept-fetch-record

### CODE REVIEW

File: Userland/Libraries/LibWeb/Fetch/Infrastructure/FetchRecord.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+#pragma once
+
+#include <LibWeb/Fetch/Infrastructure/HTTP/Requests.h>
+
+namespace Web::Fetch::Infrastructure {
+
+// https://fetch.spec.whatwg.org/#concept-fetch-record
+class FetchRecord : public JS::Cell {
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/Fetch/Infrastructure/FetchRecord.h:14]: `FetchRecord` needs `visit_edges()` for `m_request` and `m_fetch_controller`

File: Userland/Libraries/LibWeb/HTML/Scripting/Environments.h [hunk 2]
```diff
@@ -135,6 +137,10 @@ struct EnvironmentSettingsObject : public Environment {
 
     // https://html.spec.whatwg.org/multipage/webappapis.html#about-to-be-notified-rejected-promises-list
     Vector<JS::Handle<JS::Promise>> m_about_to_be_notified_rejected_promises_list;
+
+    // https://fetch.spec.whatwg.org/#concept-fetch-record
+    // A fetch group holds an ordered list of fetch records
+    Vector<JS::NonnullGCPtr<Fetch::Infrastructure::FetchRecord>> m_fetch_group;
```
Comment (kalenikaliaksandr) [inline Userland/Libraries/LibWeb/HTML/Scripting/Environments.h:143]: `m_fetch_group` needs to be visited by GC.

**Review by kalenikaliaksandr — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp [hunk 3]
```diff
@@ -1629,9 +1630,38 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<PendingResponse>> http_network_or_cache_fet
...
+            for (auto const& fetch_record : in_flight_records) {
+                // 1. Let inflightRequest be fetchRecord’s request.
+                auto const& in_flight_request = fetch_record->request();
+
+                // 2. Increment inflightKeepaliveBytes by inflightRequest’s body’s length.
+                if (in_flight_request->body().has<ByteBuffer>())
+                    inflight_keep_alive_bytes += in_flight_request->body().get<ByteBuffer>().size();
+                if (in_flight_request->body().has<JS::NonnullGCPtr<Infrastructure::Body>>() && in_flight_request->body().get<JS::NonnullGCPtr<Infrastructure::Body>>()->length().has_value()) {
+                    inflight_keep_alive_bytes += in_flight_request->body().get<JS::NonnullGCPtr<Infrastructure::Body>>()->length().value();
+                }
```
Comment (kennethmyhra) [inline Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp:1658]: We can visit this
```suggestion
                inflight_keep_alive_bytes += in_flight_request->body().visit(
                    [](Empty) { return 0ul; },
                    [](ByteBuffer const& buffer) {return buffer.size(); },
                    [](JS::NonnullGCPtr<Infrastructure::Body> body) {
                        return body->length().has_value() ? body->length().value() : 0ul;
                    });
```

**Review by kennethmyhra — COMMENTED** [review]
Added some comments on formatting changes and copyright changes that should be in the first commit.

Maybe `FetchRecord` and `fetch_group()` could each be their own commit and then the implementation of the steps that requires `fetch records` and `fetch group` concepts be in a third commit?

**Comment by AtkinsSJ** [comment]
Hi, not a comment on your code, but please click the `[ Resolve conversation ]` button on each review comment after you've pushed changes that resolve that issue. It makes it a lot clearer what is done and what isn't. :^)

**Review by kennethmyhra — COMMENTED** [review]
Looks good, just a minor nit + wrap the commit text when you exceed 72 characters, now they are arbitrary wrapped between 50-something and 60-something characters.

Also, as Sam pointed out make sure to resolve the comments you have addressed

**Review by kennethmyhra — APPROVED** [review]

**Comment by kennethmyhra** [comment]
Seems we are breaking macOS CI in the `visit()`, there's a mix of `u64` and `size_t` being returned that seems to cause the breakage. I do not have a Mac to test it out on, but try to put a `static_cast` in the `ByteBuffer` section like shown below (unless someone else comes by that has a better idea).
```cpp
[](ByteBuffer const& buffer) { return static_cast<u64>(buffer.size()); },
```

**Comment by mobounya** [comment]
macOS CI still fails, I guess someone with access to a Mac should try it. doing a force push and a full build each time seems a bit tedious

**Comment by ADKaster** [comment]
Given the error 

Users/runner/work/ladybird/ladybird/AK/Variant.h:118:13: error: 'decltype(auto)' in return type deduced as 'unsigned long long' here but deduced as 'unsigned long' in earlier return statement
            return visit<Self, Visitor, CurrentIndex + 1>(self, id, data, forward<Visitor>(visitor));
            ^

Presumably the code you added is making assumptions about the type equivalence of size_t and u64. Or some similar non-portable integer type issue.

File: Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp [hunk 4]
```diff
@@ -1629,9 +1630,39 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<PendingResponse>> http_network_or_cache_fet
...
+            for (auto const& fetch_record : in_flight_records) {
+                // 1. Let inflightRequest be fetchRecord’s request.
+                auto const& in_flight_request = fetch_record->request();
+
+                // 2. Increment inflightKeepaliveBytes by inflightRequest’s body’s length.
+                inflight_keep_alive_bytes += in_flight_request->body().visit(
+                    [](Empty) -> u64 { return 0ul; },
+                    [](ByteBuffer const& buffer) -> u64 { return buffer.size(); },
+                    [](JS::NonnullGCPtr<Infrastructure::Body> body) -> u64 {
+                        return body->length().has_value() ? body->length().value() : 0ul;
```
Comment (kennethmyhra) [inline Userland/Libraries/LibWeb/Fetch/Fetching/Fetching.cpp:1658]: Now that you are specifying the return type explicitly for the lambdas it should be safe to remove the type suffixes.

I tested this change for macOS Clang CI in my own fork: https://github.com/kennethmyhra/ladybird/actions/runs/10018680213/job/27694305833?pr=4

```suggestion
                    [](Empty) -> u64 { return 0; },
                    [](ByteBuffer const& buffer) -> u64 { return buffer.size(); },
                    [](JS::NonnullGCPtr<Infrastructure::Body> body) -> u64 {
                        return body->length().has_value() ? body->length().value() : 0;
```

**Review by kennethmyhra — APPROVED** [review]
Thanks, this looks good now!

Apologies for leading you down the path of type deducing in the lambdas.

**Comment by mobounya** [comment]
> Thanks, this looks good now!
> 
> Apologies for leading you down the path of type deducing in the lambdas.

No problem at all, thanks everyone for the reviews, happy to have my first Ladybird contribution :tada: :tada:

**Pull request merged**